In [1]:
import pandas as pd

# Direct Marketing Campaigns of Portuguese Banking Clients
2 years worth of data (2008-2010).
4,119 records.
20 independent variables.  
1 dependent variable.

**Note from the authors of the dataset:**
> Missing Attribute Values: There are several missing values in some categorical attributes, all coded with the "unknown" label. These missing values can be treated as a possible class label or using deletion or imputation techniques.

Will define the NA value when reading in the dataset.

In [2]:
banking_raw = pd.read_csv('./data/bank-original.csv', header=0, sep=";", na_values="unknown")
banking_raw.shape

(4119, 21)

In [3]:
banking_raw.head().T

0            1            2            3  \
age                      30           39           25           38   
job             blue-collar     services     services     services   
marital             married       single      married      married   
education          basic.9y  high.school  high.school     basic.9y   
default                  no           no           no           no   
housing                 yes           no          yes          NaN   
loan                     no           no           no          NaN   
contact            cellular    telephone    telephone    telephone   
month                   may          may          jun          jun   
day_of_week             fri          fri          wed          fri   
duration                487          346          227           17   
campaign                  2            4            1            3   
pdays                   999          999          999          999   
previous                  0            0            0            0   
poutcome        nonexistent  nonexistent  nonexistent  nonexistent   
emp.var.rate           -1.8          1.1          1.4          1.4   
cons.price.idx       92.893       93.994       94.465       94.465   
cons.conf.idx         -46.2        -36.4        -41.8        -41.8   
euribor3m             1.313        4.855        4.962        4.959   
nr.employed          5099.1         5191       5228.1       5228.1   
y                        no           no           no           no   

                                4  
age                            47  
job                        admin.  
marital                   married  
education       university.degree  
default                        no  
housing                       yes  
loan                           no  
contact                  cellular  
month                         nov  
day_of_week                   mon  
duration                       58  
campaign                        1  
pdays                         999  
previous                        0  
poutcome              nonexistent  
emp.var.rate                 -0.1  
cons.price.idx               93.2  
cons.conf.idx                 -42  
euribor3m                   4.191  
nr.employed                5195.8  
y                              no

In [20]:
banking_raw.describe(include="all")

age     job  marital          education default housing  loan  \
count   4119.000000    4080     4108               3952    3316    4014  4014   
unique          NaN      11        3                  7       2       2     2   
top             NaN  admin.  married  university.degree      no     yes    no   
freq            NaN    1012     2509               1264    3315    2175  3349   
mean      40.113620     NaN      NaN                NaN     NaN     NaN   NaN   
std       10.313362     NaN      NaN                NaN     NaN     NaN   NaN   
min       18.000000     NaN      NaN                NaN     NaN     NaN   NaN   
25%       32.000000     NaN      NaN                NaN     NaN     NaN   NaN   
50%       38.000000     NaN      NaN                NaN     NaN     NaN   NaN   
75%       47.000000     NaN      NaN                NaN     NaN     NaN   NaN   
max       88.000000     NaN      NaN                NaN     NaN     NaN   NaN   

         contact month day_of_week     duration     campaign        pdays  \
count       4119  4119        4119  4119.000000  4119.000000  4119.000000   
unique         2    10           5          NaN          NaN          NaN   
top     cellular   may         thu          NaN          NaN          NaN   
freq        2652  1378         860          NaN          NaN          NaN   
mean         NaN   NaN         NaN   256.788055     2.537266   960.422190   
std          NaN   NaN         NaN   254.703736     2.568159   191.922786   
min          NaN   NaN         NaN     0.000000     1.000000     0.000000   
25%          NaN   NaN         NaN   103.000000     1.000000   999.000000   
50%          NaN   NaN         NaN   181.000000     2.000000   999.000000   
75%          NaN   NaN         NaN   317.000000     3.000000   999.000000   
max          NaN   NaN         NaN  3643.000000    35.000000   999.000000   

           previous     poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  \
count   4119.000000         4119   4119.000000     4119.000000    4119.000000   
unique          NaN            3           NaN             NaN            NaN   
top             NaN  nonexistent           NaN             NaN            NaN   
freq            NaN         3523           NaN             NaN            NaN   
mean       0.190337          NaN      0.084972       93.579704     -40.499102   
std        0.541788          NaN      1.563114        0.579349       4.594578   
min        0.000000          NaN     -3.400000       92.201000     -50.800000   
25%        0.000000          NaN     -1.800000       93.075000     -42.700000   
50%        0.000000          NaN      1.100000       93.749000     -41.800000   
75%        0.000000          NaN      1.400000       93.994000     -36.400000   
max        6.000000          NaN      1.400000       94.767000     -26.900000   

          euribor3m  nr.employed     y  
count   4119.000000  4119.000000  4119  
unique          NaN          NaN     2  
top             NaN          NaN    no  
freq            NaN          NaN  3668  
mean       3.621356  5166.481695   NaN  
std        1.733591    73.667904   NaN  
min        0.635000  4963.600000   NaN  
25%        1.334000  5099.100000   NaN  
50%        4.857000  5191.000000   NaN  
75%        4.961000  5228.100000   NaN  
max        5.045000  5228.100000   NaN

*Please note, according to the dataset description, a client that had not been contacted previously is represented with the value 999 in the "contact" column.*

In [ ]:
banking_raw.isna().sum()

Will drop all records that contain any null values since we need to reduce our dataset to <= 3,000 records.

In [3]:
banking_nonulls = banking_raw.dropna(how='any')
banking_nonulls.shape

(3090, 21)

In [4]:
bank = banking_nonulls

In summary, we:
- imported our original 4,119 record dataset
- defined & identified the null values present
- removed the rows with any null value present

We will keep the dataset as is (3,090) as permitted by the professor, and will export `banking_nonulls` dataframe to a CSV.

In [5]:
bank.rename(columns={"emp.var.rate": "emp_var_rate", "cons.price.idx": "cons_price_idx", "cons.conf.idx": "cons_conf_idx", "nr.employed": "nr_employed"}, 
            inplace=True)


/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [23]:
bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx',
       'cons_conf_idx', 'euribor3m', 'nr_employed', 'y'],
      dtype='object')

In [6]:
bank.reset_index(inplace=True, drop=True)

In [7]:
bank.head(10)

age            job   marital            education default housing loan  \
0   30    blue-collar   married             basic.9y      no     yes   no   
1   39       services    single          high.school      no      no   no   
2   25       services   married          high.school      no     yes   no   
3   47         admin.   married    university.degree      no     yes   no   
4   32       services    single    university.degree      no      no   no   
5   32         admin.    single    university.degree      no     yes   no   
6   31       services  divorced  professional.course      no      no   no   
7   36  self-employed    single             basic.4y      no      no   no   
8   36         admin.   married          high.school      no      no   no   
9   47    blue-collar   married             basic.4y      no     yes   no   

     contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0   cellular   may         fri  ...         2    999         0  nonexistent   
1  telephone   may         fri  ...         4    999         0  nonexistent   
2  telephone   jun         wed  ...         1    999         0  nonexistent   
3   cellular   nov         mon  ...         1    999         0  nonexistent   
4   cellular   sep         thu  ...         3    999         2      failure   
5   cellular   sep         mon  ...         4    999         0  nonexistent   
6   cellular   nov         tue  ...         1    999         1      failure   
7   cellular   jul         thu  ...         1    999         0  nonexistent   
8  telephone   may         wed  ...         2    999         0  nonexistent   
9  telephone   jun         thu  ...         2    999         0  nonexistent   

  emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  nr_employed   y  
0         -1.8          92.893          -46.2      1.313       5099.1  no  
1          1.1          93.994          -36.4      4.855       5191.0  no  
2          1.4          94.465          -41.8      4.962       5228.1  no  
3         -0.1          93.200          -42.0      4.191       5195.8  no  
4         -1.1          94.199          -37.5      0.884       4963.6  no  
5         -1.1          94.199          -37.5      0.879       4963.6  no  
6         -0.1          93.200          -42.0      4.153       5195.8  no  
7          1.4          93.918          -42.7      4.968       5228.1  no  
8          1.1          93.994          -36.4      4.859       5191.0  no  
9          1.4          94.465          -41.8      4.958       5228.1  no  

[10 rows x 21 columns]

In [32]:
# bank.to_csv('./data/cleaned-bank.csv')

<font color="red">**NOTE:** Add `i` as the first column name to indicate index. </font> Can do so on the terminal by inputing:
   - `vi cleaned-bank.csv` to edit the file
   - `i` to insert text
   - On first line, enter `i` right before the comma.
   - `esc`
   - `:wq` to write changes and quit 
    
Need to do this for SAS code to work.

# Actual Analysis of Cleaned Dataset

In [8]:
pd.set_option('display.max_columns', None)

In [9]:
bank.describe(include="all").T

count unique                top  freq       mean       std  \
age             3090    NaN                NaN   NaN    39.1841    10.184   
job             3090     11             admin.   854        NaN       NaN   
marital         3090      3            married  1791        NaN       NaN   
education       3090      7  university.degree  1107        NaN       NaN   
default         3090      2                 no  3089        NaN       NaN   
housing         3090      2                yes  1688        NaN       NaN   
loan            3090      2                 no  2583        NaN       NaN   
contact         3090      2           cellular  2108        NaN       NaN   
month           3090     10                may   981        NaN       NaN   
day_of_week     3090      5                mon   642        NaN       NaN   
duration        3090    NaN                NaN   NaN    259.245    263.96   
campaign        3090    NaN                NaN   NaN    2.50939   2.59817   
pdays           3090    NaN                NaN   NaN    953.351    208.03   
previous        3090    NaN                NaN   NaN   0.208091  0.560141   
poutcome        3090      3        nonexistent  2602        NaN       NaN   
emp_var_rate    3090    NaN                NaN   NaN -0.0467961   1.59628   
cons_price_idx  3090    NaN                NaN   NaN    93.5322  0.585886   
cons_conf_idx   3090    NaN                NaN   NaN   -40.6186   4.74645   
euribor3m       3090    NaN                NaN   NaN    3.48171   1.77002   
nr_employed     3090    NaN                NaN   NaN    5161.14   76.4822   
y               3090      2                 no  2720        NaN       NaN   

                   min     25%     50%     75%     max  
age                 20      31      37      46      88  
job                NaN     NaN     NaN     NaN     NaN  
marital            NaN     NaN     NaN     NaN     NaN  
education          NaN     NaN     NaN     NaN     NaN  
default            NaN     NaN     NaN     NaN     NaN  
housing            NaN     NaN     NaN     NaN     NaN  
loan               NaN     NaN     NaN     NaN     NaN  
contact            NaN     NaN     NaN     NaN     NaN  
month              NaN     NaN     NaN     NaN     NaN  
day_of_week        NaN     NaN     NaN     NaN     NaN  
duration             0     104     181     315    3643  
campaign             1       1       2       3      35  
pdays                0     999     999     999     999  
previous             0       0       0       0       6  
poutcome           NaN     NaN     NaN     NaN     NaN  
emp_var_rate      -3.4    -1.8     1.1     1.4     1.4  
cons_price_idx  92.201  93.075  93.444  93.994  94.767  
cons_conf_idx    -50.8   -42.7   -41.8   -36.4   -26.9  
euribor3m        0.635   1.313   4.856   4.961   5.045  
nr_employed     4963.6  5099.1    5191  5228.1  5228.1  
y                  NaN     NaN     NaN     NaN     NaN

**pdays**
-	Days passed since previous campaign’s contact.
-	A client that had not been contacted previously is represented with the value 999.

I interpret a value of 999 as almost equivalent to a N/A value, and a value of 0 as the client being contacted the same day?  Hopefully our model doesn't deem this attribute significant so we don't have to figure out what to do with these value discrepancies (see unique value frequency counts below). 

In [16]:
# frequency/count of each unique value
bank['pdays'].value_counts()

999    2948
3        46
6        40
4        13
10        8
7         6
12        5
2         4
5         4
1         3
0         2
15        2
9         2
13        2
18        1
16        1
21        1
11        1
19        1
Name: pdays, dtype: int64

In [17]:
bank['month'].value_counts()

may    981
jul    514
aug    495
nov    387
jun    365
apr    169
oct     61
sep     57
mar     42
dec     19
Name: month, dtype: int64

Month5 (May) was seen as significant in the STEPWISE selection method.  This is biased because it's the month with the most calls.  
Month3 (March) followed as significant.  It has the 2nd lowest frequency (good thing), but it's also the 1st month of the campaign. 
Finally, month6 (June) followed as significant. This month may also be biased because it has the 2nd highest frequency of calls.

The selection method chose 8 out of the 53 attributes, but the R2 value was only 0.2637.


We have:
- 2720 samples of 'no' => 88% Probability,
- and 370 samples of 'yes' => 12% Probability.

odds(y=1) = 0.12/0.88 => The odds that event Y = 1 occurs is 0.136 to 1.  This means we have a higher chance of failure.

odds(y=0) = 0.88/0.12 => 7.33  to 1 	(slide 8 of lecture 7)

I want to test if removing some observations would increase the Rsquare value, specifically observations that meet the following criteria:
   - target = 0
   - pdays = 999 (b/c 95% of our data is comprised of those values)
   - month = 'may' (month w/the highest amount of calls)

Another option could be to add in more 'yes' samples from the full dataset to add more variance and increase the probability that event Y will occur.
